parsons/action_builder/action_builder.py

```diff
@@ -0,0 +1,351 @@
+        tags in excess.
+        `Args:`
+            identifiers: list
+                The list of strings of unique identifiers for a record being updated. ID strings
+                will need to begin with the origin system, followed by a colon, e.g.
+                `action_builder:abc123-...`.
+            tag_name: str or list
+                The name(s) of the new tag(s), i.e. the custom field value(s).
+            tag_field: str or list
+                The name(s) of the tag category(ies), i.e. the custom field name(s). Must have
+                equal or fewer items as `tag_name`.
+            tag_section: str or list
+                The name(s) of the tag section(s), i.e. the custom field group name(s). Must have
+                equal or fewer items as `tag_section`.
+            campaign: str
+                Optional. The 36-character "interact ID" of the campaign whose data is to be
+                retrieved or edited. Not necessary if supplied when instantiating the class.
+        `Returns:`
+            Dict containing Action Builder entity data of the entity being tagged.
+        """
+
+        # Ensure all tag args are lists
+        tag_name = tag_name if isinstance(tag_name, list) else [tag_name]
+        tag_field = tag_field if isinstance(tag_field, list) else [tag_field]
+        tag_section = tag_section if isinstance(tag_section, list) else [tag_section]
+
+        # Use lists of tuples to identify length ordering in case amounts differ
+        lengths = []
+        lengths.append(
+            ("name", len(tag_name))
+        )  # Most specific - should be >= other lengths
+        lengths.append(("field", len(tag_field)))
+        lengths.append(
+            ("section", len(tag_section))
+        )  # Most generic - should be <= other lengths
+
+        # Sort by length, maintaining original order in ties
+        ordered_lengths = sorted(lengths, key=lambda x: x[1], reverse=True)
+        sorted_keys = [x[0] for x in ordered_lengths]
+
+        # Raise an error if there are fewer specific items provided than generic
+        if sorted_keys[0] != "name":
+            raise ValueError(
+                "Not enough tag_names provided for tag_fields or tag_sections"
+            )
+
+        if sorted_keys[1] != "field":
+            raise ValueError("Not enough tag_fields provided for tag_sections")
+
+        # Construct tag data, applying final generic taxonomy to any excess specific items
+        tag_data = [
+            {
+                "action_builder:name": x,
+                "action_builder:field": tag_field[min(i, len(tag_field) - 1)],
+                "action_builder:section": tag_section[min(i, len(tag_section) - 1)],
+            }
+            for i, x in enumerate(tag_name)
+        ]
+
+        data = {"add_tags": tag_data}
+
+        return self.upsert_entity(identifiers=identifiers, data=data, campaign=campaign)
+
+    def upsert_connection(self, identifiers, tag_data=None, campaign=None):
```

Comment:
unless there's an insane number of entities and therefore types of connections, would also lean towards having a more explicit public api of like `upsert_person_campaign_connection` and using this as a private method to keep things dry

Reply:
I'm not sure I completely understand the suggestion here. Just in case there's any misunderstanding about connections and how they're applied, I'll try to summarize here.

Connections can exist between any two entity records, regardless of whether the entity type is person or not. The system is strict about there only ever being one connection record between two specific entity records, and the connection helper endpoint is pretty much the only way to create or update connections (other than the signup helper, which doesn't really work differently in this regard). So you don't actually ever know for sure (unless you make wasteful additional API calls) whether you're inserting or updating.

Connection records can accept tags, much the same way as entity records. These can help illuminate important information about the relationship between entities. For instance, between two people records, you could have a tag indicate the human relationship between them, or on a connection between a person and an organization, you could have one that indicates the person's role at that organization.

Given all this, I have a hard time imagining what public method we could write that would be more intuitive than this one, and use this one as internal. But very open to anything I'm just not seeing!